README.md
# enchilada [![Build Status](https://secure.travis-ci.org/defunctzombie/node-enchilada.png?branch=master)](http://travis-ci.org/defunctzombie/node-enchilada)

serve up your javascript files all wrapped up using [browserify](https://github.com/substack/node-browserify). Yum!

## with express/connect

```javascript
var app = express();

// serves up all your javascript files, handling all require() calls
app.use(enchilada(__dirname + '/public'));

// fallback for other static resources
app.use(express.static(__dirname + '/public'));
```

Now just visit any ```.js``` url which maps to a path under /public and the packaged file will be served.

## sourcemaps

Versions 0.7+ of enchilada do not bundle sourcemaps with your javascript files; instead sourceMapURL comment is used. This allows for specifying both the `compress` and `debug` options as true in production without impacting users but still benefiting from being able to obtain sourcemaps. Care should be taken to limit access to the mapfiles (served via the /path/to/original/js/script.map.json) usually with middleware before enchilada.

## options

No one likes a stale enchilada. Out in the real world, you want to leverage browser caching for rarely changing files. Imagine that your project uses files like jquery or [engine.io](https://github.com/LearnBoost/engine.io-client), these files don't change as much as your app code. It would be silly to keep sending them with every js file you serve up. Enchilada makes this easy to do.

Just add the proper ingredients and your enchilada will be served up as you requested.

```javascript
app.use(enchilada({
    src: __dirname + '/public', // location of your js files
    cache: true || false, // default false (use true for production to disable file watching)
    compress: true || false, // default false
    debug: true || false, // default false (enable sourcemap output with bundle)
    watchCallback: function(filename) {}, // optional (use to do something clever, like tell client to reload the page)
    routes: {
        // key is the url route, value is either a file relative to src
        '/js/jquery.js': './js/jquery.js',
        // or a module installed via npm
        '/js/engine.io.js': 'engine.io-client'
    },
    transforms: [ handleify, brfs ]
}));
```

Now just make sure you load the required scripts before any other js file that might use them.

```html
<!-- load the scripts we know will be used by several files -->
<script src="/js/jquery.js"></script>
<script src="/js/engine.io.js"></script>

<!-- load other js files as you would before -->
<script src="/js/app.js"></script>
```

## examples

See the [examples](examples) directory for working code you can copy and paste.

## install

Install with [npm](https://npmjs.org)

```shell
npm install enchilada
```


Changelog
unreleased

0.7.1 - 2014/1/4
	* sourceMappingUrl is relative file path

0.7.0 - 2014/1/3
	* Strip sourcemaps from js files and server under separate map url
	* update browserify to 3.x line 3.18.0

0.6.0
	* Support for regular connect and regular http servers

0.5.1
	* update browserify to 2.36.0

0.5.0
	* bundles from "routes" are now also entry files and run automatically
	when loaded on the page


test/basic.js
var assert = require('assert');
var vm = require('vm');

var request = require('supertest');
var express = require('express');
var after = require('after');

var enchilada = require('../');

var app = express();

suite('basic');

test('setup', function(done) {
    app.use(enchilada(__dirname + '/assets'));
    app.listen(done);
});

test('basic', function(done) {
    done = after(3, done);

    request(app)
    .get('/foo.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 200);
        assert.equal(res.headers['content-type'], 'application/javascript');

        var sandbox = {
            done: done,
            assert: assert
        };

        var src = res.text;
        vm.runInNewContext(src, sandbox);
        done();
    });
});



test/routes.js
var assert = require('assert');
var vm = require('vm');

var request = require('supertest');
var express = require('express');
var after = require('after');

var enchilada = require('../');

var app = express();

suite('routes');

test('setup', function(done) {
    app.use(enchilada({
        src: __dirname + '/assets',
        routes: {
            '/foo.js': './foo.js',
            '/cats-module.js': 'cats-module'
        }
    }));
    app.listen(done);
});

test('/foo.js - should run', function(done) {
    done = after(3, done);

    request(app)
    .get('/foo.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 200);
        assert.equal(res.headers['content-type'], 'application/javascript');

        var sandbox = {
            done: done,
            assert: assert
        };

        var src = res.text;
        vm.runInNewContext(src, sandbox);
        done();
    });
});

test('/baz.js - entry', function(done) {

    // done is called inside every module file
    // and once at the end of response
    done = after(5, done);

    request(app)
    .get('/foo.js')
    .end(function(err, res) {

        var sandbox = {
            done: done,
            assert: assert
        };

        // inject foo into scope
        var src = res.text;
        vm.runInNewContext(src, sandbox);

        // load baz and run that too
        request(app)
        .get('/baz.js')
        .end(function(err, res) {
            assert.equal(res.statusCode, 200);
            assert.equal(res.headers['content-type'], 'application/javascript');

            var src = res.text;
            vm.runInNewContext(src, sandbox);

            // run again, the entry will be re-run, but if foo.js is proper external
            // it will not be re-run (thus done will be called only 4 times)
            vm.runInNewContext(src, sandbox);

            done();
        });
    });
});

test('/cats-module.js - should run', function(done) {

    done = after(2, done);

    request(app)
    .get('/cats-module.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 200);
        assert.equal(res.headers['content-type'], 'application/javascript');

        var sandbox = {
            done: done,
            assert: assert
        };

        var src = res.text;
        vm.runInNewContext(src, sandbox);
        done();
    });
});

test('/cats.js - entry', function(done) {
    done = after(4, done);

    request(app)
    .get('/cats-module.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 200);
        assert.equal(res.headers['content-type'], 'application/javascript');

        var sandbox = {
            done: done,
            assert: assert
        };

        var src = res.text;
        vm.runInNewContext(src, sandbox);

        request(app)
        .get('/cats.js')
        .end(function(err, res) {
            assert.equal(res.statusCode, 200);
            assert.equal(res.headers['content-type'], 'application/javascript');

            var src = res.text;
            vm.runInNewContext(src, sandbox);

            // and again to make sure we did not include the module source
            vm.runInNewContext(src, sandbox);
            done();
        });
    });
});


test/server.js
var assert = require('assert');
var vm = require('vm');

var request = require('supertest');
var after = require('after');
var http = require('http');

var enchilada = require('../');

suite('server');

var server;

test('setup normal http server', function(done) {
    server = http.createServer(enchilada(__dirname + '/assets'));
    server.listen(done);
});

test('basic without express', function(done) {
    done = after(3, done);

    request(server)
    .get('/foo.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 200);
        assert.equal(res.headers['content-type'], 'application/javascript');

        var sandbox = {
            done: done,
            assert: assert
        };

        var src = res.text;
        vm.runInNewContext(src, sandbox);
        done();
    });
});

test('handle browserify errors', function(done) {
    request(server)
    .get('/error.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 500);
        assert.equal(res.headers['content-type'], 'text/plain');
        assert.ok(/Error/.test(res.text));
        done();
    });
});

test('handle missing file', function(done) {
    request(server)
    .get('/missing.js')
    .end(function(err, res) {
        assert.equal(res.statusCode, 404);
        assert.equal(res.headers['content-type'], 'text/plain');
        assert.equal(res.text, 'Not Found');
        done();
    });
});


watcher.js
var fs = require('fs')
var path = require('path')
var EventEmitter = require('events').EventEmitter
var debug = require('debug')('enchilada:watch');

var emitter = new EventEmitter
var listeners = {}

module.exports = function onChange(filename, callback) {
    filename = path.normalize(filename)

    // First listener on filename
    if (!listeners[filename]) {

        listeners[filename] = function(curr, prev) {
            if (!curr || !prev || curr.mtime !== prev.mtime) {
                emitter.emit(filename, curr, prev)
            }
        }

        debug('Watching %s', filename)

        // Watch the file
        fs.watchFile(filename, {
            persistent:false,
            interval: 1007
        }, listeners[filename])
    }

    // Bind callback for changes on file
    emitter.on(filename, callback)

    // Pseudo-FS.watcher interface
    return {
        close: function() {
            emitter.removeListener(filename, callback)
            if (emitter.listeners(filename).length === 0) {
                debug('Unwatching %s', filename)
                fs.unwatchFile(filename, listeners[filename])
                delete listeners[filename]
            }
        }
    }
}


index.js
var path = require('path');
var fs = require('fs');
var url = require('url');
var crypto = require('crypto');
var convert = require('convert-source-map');

var mime = require('mime');
var uglifyjs = require('uglify-js');
var browserify = require('browserify');
var debug = require('debug')('enchilada');

var watcher = require('./watcher')

module.exports = function enchilada(opt) {

    // if just a path is passed in, treat as public file dir
    if (typeof opt === 'string') {
        opt = { src: opt };
    }

    var pubdir = opt.src;
    var routes = opt.routes || {};
    var bundles = {};

    var compress = false || opt.compress;
    var cache = {}; // cache of sourcefiles
    var maps = {}; // cache of sourcemaps
    var debug_opt = false || opt.debug;

    var watch = !opt.cache;
    var watchCallback = opt.watchCallback;

    function makeBundle(options) {
        var bundle = browserify(options);
        if (opt.transforms) {
            opt.transforms.forEach(function(transform) {
                bundle.transform(transform)
            });
        }
        if (opt.externals) {
            opt.externals.forEach(function(external) {
                bundle.external(external);
            });
        }
        return bundle;
    }

    // TODO(shtylman) externs that use other externs?
    Object.keys(routes).map(function(id) {
        var name = routes[id];

        debug('route: %s -> %s', id, name);

        var bundle = makeBundle({ exposeAll: true });
        bundle.require(name, { entry: true, expose: name, basedir: pubdir });
        return bundles[id] = bundle;
    });

    return function(req, res, next) {
        var req_path = req.path || url.parse(req.url).path;

        if (/.map.json$/.test(req_path) && maps[req_path]) {
            return res.json(maps[req_path]);
        }

        // if no extension, then don't process
        // handles case of directories and other random urls
        if (!path.extname(req_path)) {
            return notFound();
        }
        else if (mime.lookup(req_path) !== 'application/javascript') {
            return notFound();
        }

        // check cache
        var cached = cache[req_path];
        if (cached) {
            return sendResponse(null, cached);
        }

        // check for bundle
        var bundle = bundles[req_path];
        if (bundle) {
            return generate(bundle, sendResponse);
        }

        var local_file = path.normalize(path.join(pubdir, req_path));

        // check for malicious attempts to access outside of pubdir
        if (local_file.indexOf(pubdir) !== 0) {
            return notFound();
        }

        debug('bundling %s', local_file);

        // lookup in filesystem
        fs.exists(local_file, function(exists) {
            if (!exists) {
                return notFound();
            }

            var bundle = makeBundle(local_file);
            Object.keys(bundles).forEach(function(id) {
                bundle.external(bundles[id]);
            });
            generate(bundle, sendResponse);
        });

        function generate(bundle, callback) {
            var dependencies = {};

            // typically SyntaxError
            var otherError;
            bundle.once('error', function(err) { otherError = err; });

            var collect_deps = function(file) {
                dependencies[file] = true;
            };

            if (watch) {
                bundle.on('file', collect_deps);
            }

            bundle.bundle({ debug: debug_opt }, function(err, src) {
                bundle.removeListener('file', collect_deps);

                if (watch) {
                    watchFiles(bundle, dependencies, req_path);
                }
                if (err) {
                    return callback(err);
                }
                if (otherError) {
                    return callback(otherError);
                }

                var srcmap = undefined;
                var map_path = undefined;
                if (debug_opt) {
                    // output sourcemap
                    srcmap = convert.fromComment(src);
                    src = convert.removeComments(src);
                    srcmap.setProperty('file', req_path);
                    map_path = req_path.replace(/.js$/, '.map.json');
                }

                if (compress) {
                    var ugly_opt = {
                        fromString: true
                    };

                    if (srcmap) {
                        ugly_opt.inSourceMap = srcmap.toObject(),
                        ugly_opt.outSourceMap = req_path
                    }

                    var result = uglifyjs.minify(src, ugly_opt);

                    src = result.code;

                    if (srcmap) {
                        // prepare new sourcemap
                        // we need to get the sources from bundled sources
                        // uglify does not carry those through
                        var srcs = srcmap.getProperty('sourcesContent');
                        srcmap = convert.fromJSON(result.map);
                        srcmap.setProperty('sourcesContent', srcs);
                    }
                }

                if (srcmap) {
                    src += '//# sourceMappingURL=' + path.basename(map_path);
                    maps[map_path] = srcmap.toObject();
                }

                cache[req_path] = src;

                callback(null, src);
            });
        }

        function notFound() {
            if (typeof next === 'function') {
                return next();
            }

            res.setHeader('Content-Type', 'text/plain');
            respond(404, 'Not Found');
        }

        function sendResponse(err, src) {
            if (err) {
                return sendError(err);
            }

            res.setHeader('Content-Type', 'application/javascript');
            res.setHeader('ETag', crypto.createHash('md5').update(src).digest('hex').slice(0, 6));
            res.setHeader('Vary', 'Accept-Encoding');
            respond(200, src);
        }

        function sendError(err) {
            if (typeof next === 'function') {
                return next(err);
            }

            res.setHeader('Content-Type', 'text/plain');
            respond(500, err.toString());
        }

        function respond(code, data) {
            data = new Buffer(data);
            res.setHeader('Content-Length', data.length);
            res.writeHeader(code);
            res.end(data);
        }

        function watchFiles(bundle, dependencies, path) {
            var watchers = Object.keys(dependencies).map(function(filename) {
                return watcher(filename, function() {
                    delete cache[path];
                    watchers.forEach(function(watcher) {
                        watcher.close();
                    });
                    generate(bundle, function(error) {
                        watchCallback && watchCallback(error, path);
                    });
                });
            });
        }
    };
};

